text/042-unify-definition-fixpoint-syntax.md

```diff
@@ -0,0 +1,305 @@
+
+# Drawbacks
+
+- This creates more ways of doing the same thing.  People could decide
+  to write things like:
+
+  ```coq
+  #[ unsealed ]
+  Theorem negb (b : bool) :=
+    match b with true => false | false => true end.
+
+  #[ sealed ]
+  Definition negb_idem (b : bool) : negb (negb b) = b :=
+    match b with true => eq_refl | false => eq_refl end.
+  ```
+
+  But it seems consistent with the current philosophy of the language
+  to give many synonyms for doing the same thing and trusting the user
+  to produce readable scripts in the end.
+
+- The first part of the CEP will impose users of the existing `Theorem
+  ... with` syntax to add a `recursive` attribute to their
+  declarations.
+
+- The second part of the CEP will impose users of the `Theorem` /
+  `Defined` combo to add an `unsealed` attribute to their declarations
+  or to set some of the new flags.  Thanks to the flags, this porting
+  step should be pretty uninvasive.
+
+# Alternatives
+
+- We could keep the syntax separate and keep waiting for users asking
+  for this or that extension to implement it, but it is unlikely that
+  the syntax would ever converge, and it would remain inconsistent and
+  confusing.  The first chapter of the refactored reference manual
+  could rely on the unsatisfactory `Proof @term` syntax to avoid
+  bringing the tactics in.
+
+- We could implement this plan without the second part.  This would
+  not be a big problem, but it would leave a small inconsistency
+  between the way transparency is set when the body is provided or
+  not.  In this case, the `sealed` and `unsealed` attributes would
+  raise an error when the body is not provided.
+
+- We could implement this plan without providing the `sealed` and
+  `unsealed` attributes at all.  This means that when a body is
+  provided, `Theorem` and `Definition` would not be interchangeable
+  (but they already mostly are when in proof mode because the `Qed`
+  and `Defined` commands are not forced upon the user, so this would
+  not be consistent).
+
+- We could also decide to implement this plan but make `Definition`
+  and `Theorem` fully interchangeable.  A `Theorem` with a body would
+  be transparent by default, a `sealed` attribute could be used on
+  `Definition` or `Theorem` to make the body opaque.  In this case, we
+  would not implement the second part of the CEP either.
+
+# Unresolved questions
+
+- The current `Theorem ... with` syntax has been criticized for
+  allowing proofs that are rejected at `Qed` time (unless the
+  `Guarded` command is used).  The problem is exactly the same as when
+  using the `fix` or `cofix` tactics.  When using `Fixpoint` in proof
+  mode, it would seem that the `@fixannot` clause is ignored.  We
+  could look into making this clause being used and the use of tactic
+  becoming safer, but IMHO as a first step it suffices to properly
+  document this behavior and possibly warn when using the `@fixannot`
+  clause that it is ignored.
```

Comment:
As far as I know, the clause isn't ignored: it's only applied during `Qed`.

It's not easy to check the guard condition on partial terms: they're partial term, they are not really guarded yet.  You could have a partial rejection algorithm. But you have to be careful for it to be somewhat incremental, rather than checking the entire term every time a refinement occurs.

At any rate, this doesn't look particularly related to the current proposal. I'd just remove this paragraph.

Reply:
I put it there because @maximedenes was wary of extending further a syntax that he doesn't like.

Reply:
The problem of this syntax is not described accurately. What I complained about is the syntax *without* the clause.

Essentially, it allows to say "I'll prove this by induction" without saying induction on what (which is not exactly good mathematical practice btw). But this is not supported at the kernel level, so the system generates an ill-typed term. You can observe it by the fact that `Guarded` will fail at any step in the proof. Then, once you reach `Qed`, another hack kicks in to massage the term and finally guess what the induction was about, before sending the final term to the kernel.

Bottom line: this syntax (`Lemma ... with` without a `@fixannot` clause) reflects a mathematically dubious notion ("induction on something"), that is not supported by the kernel. The symptoms are dual to the ones you mention (`Guarded` fails everywhere, proof terms are ill-typed, but `Qed` succeeds magically).

Reply:
`@fixannot` is `{ struct ident }`, right? That's very much not ignored, and it can greatly decrease `Qed` time on big theorems (here, from 8 to 3 seconds) — so much so, my fundamental theorem is a `Fixpoint... {struct }.` (Writing a custom induction principle can be much more expensive).

It could be good to have a warning for the unsuspecting users running into this for the first time, as long as one can disable it :-)

Reply:
> You can observe it by the fact that `Guarded` will fail at any step in the proof.

IME, no; it only fails once you use an assumption on something that isn't a subterm. When I write those proofs, I regularly use `Guarded` (just checked on 8.10, also experienced on 8.9 and probably 8.11). My experience was FWIW on mutual fixpoints.

> Essentially, it allows to say "I'll prove this by induction" without saying induction on what (which is not exactly good mathematical practice btw).

That hasn't stopped Agda from having only this syntax (tho they don't do inlining, so it's cheaper for them). In theory, there is research on cyclic proofs dealing with this.
But in practice, this is just a way to do inductive proofs even when Coq doesn't generate the right induction principles (and for whatever reason you don't want to write your own, or it outright doesn't make sense).

Reply:
> let's spread this syntax to more constructions only once 

"Spread this syntax" doesn't seem a good summary. Right now, all constructs that allow recursion also allow omitting fixannot, regardless of whether they require the proof mode or not — in fact, `Theorem` and variants *forbid it*, which this CEP fixes.

The CEP does add support for recursion to Definition and Example — and allows omitting fixannot there too; so you're arguing to require fixannot there, keeping things irregular.

> we've figured out the mathematical meaning and proper kernel support for it.

I must miss something here too? I'm confident the mathematical meaning is just to infer the annotation.

Your example is clearly a valid bug report (where's the ticket?), and it sounds like the in-progress term should use `{struct ?E}` instead of `{struct y}` where `?E` is a brand-new evar. I feel that should not reach the kernel, but that's because IMHO "kernel" should be reserved for Coq's trusted core, where evars are not welcome (but I guess that's not how things are implemented).

But this CEP is orthogonal. Its point (which I agree with) is that the choice of `thm_token` is basically a required comment whose meaning is only informal — but apparently, quite a few people care about those comments (as argued well by Hugo Herbelin and Xavier Leroy in https://github.com/coq/coq/pull/8265).

Currently Coq assumes that the informal meaning is associated with a different subset of features, but this association is designed in pretty inconsistent way, and this CEP recognizes that this association doesn't always work. Plus, I suspect the implementations are inconsistent — I might or might not have observed `thm_token`-depending bugs.

Reply:
> and allows omitting fixannot there too; so you're arguing to require fixannot there, keeping things irregular.

Yes. I have a problem with arguments of uniformity with broken features. Note that the "feature" I showed above is also currently incompatible with async proofs (it should fail under CoqIDE, although I haven't tested).

> Your example is clearly a valid bug report (where's the ticket?), and it sounds like the in-progress term should use `{struct ?E}` instead of `{struct y}` where `?E` is a brand-new evar. I feel that should not reach the kernel, but that's because IMHO "kernel" should be reserved for Coq's trusted core, where evars are not welcome (but I guess that's not how things are implemented).

Evars designate subterms. Writing `{struct ?E}` makes no sense, at least in the current system. `{struct` expects an integer which designates one of the variables bound by the `fix`. For this notation to makes sense, you'd need to define this new notion of existential variable, and the guard condition when this variable still hasn't been instantiated.

Reply:
Yes, by "brand-new evar" I meant what you describe.